core/common/src/main/java/alluxio/retry/TimeBoundedRetry.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.retry;
+
+import alluxio.time.Sleeper;
+import alluxio.time.TimeContext;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+
+/**
+ * Retry mechanism which performs retries until a certain period of time has elapsed. Subclasses
+ * determine the interval between retries.
+ */
+public abstract class TimeBoundedRetry implements RetryPolicy {
+  private final Clock mClock;
+  private final Sleeper mSleeper;
+  private final Duration mMaxDuration;
+  private final Instant mStartTime;
+  private final Instant mEndTime;
+
+  private int mRetryCount = 0;
+  private boolean mDone = false;
+
+  /**
+   * @param timeCtx the time context to use for time-based operations
+   * @param maxDuration the maximum duration
+   */
+  public TimeBoundedRetry(TimeContext timeCtx, Duration maxDuration) {
+    mClock = timeCtx.clock();
+    mSleeper = timeCtx.sleeper();
+    mMaxDuration = maxDuration;
+    mRetryCount = 0;
+    mStartTime = mClock.instant();
+    mEndTime = mStartTime.plus(mMaxDuration);
+  }
+
+  @Override
+  public int getRetryCount() {
+    return mRetryCount;
+  }
+
+  @Override
+  public boolean attemptRetry() {
+    if (mDone) {
+      return false;
+    }
+    if (!mClock.instant().isBefore(mEndTime)) {
+      mDone = true;
+      return false;
+    }
+    mRetryCount++;
+    try {
+      Duration nextWaitTime = nextWaitTime();
+      Instant now = mClock.instant();
```

Comment:
Could we define this in the beginning of this method and reuse it, there doesn't seem to be anything which would take a substantial amount of time, or is this to get around the uncertain implementation in `nextWaitTime`?